src/ripple/app/misc/NegativeUNLVote.h

```diff
@@ -0,0 +1,193 @@
+     * So we skip voting for disabling them for 2 flag ledgers.
+     */
+    static constexpr size_t newValidatorDisableSkip = 256 * 2;
+    /**
+     * We only want to put 25% of the UNL on the NegativeUNL.
+     */
+    static constexpr float nUnlMaxListed = 0.25;
+
+    /**
+     * Constructor
+     *
+     * @param myId the NodeID of the local node
+     * @param j log
+     */
+    NegativeUNLVote(NodeID const& myId, beast::Journal j);
+    ~NegativeUNLVote() = default;
+
+    /**
+     * Cast our local vote on the NegativeUNL candidates.
+     *
+     * @param prevLedger the parent ledger
+     * @param unlKeys the trusted master keys of validators in the UNL
+     * @param validations the validation message container
+     * @param initialSet the transactions set for adding ttUNL_MODIFY Tx if any
+     */
+    void
+    doVoting(
+        std::shared_ptr<Ledger const> const& prevLedger,
+        hash_set<PublicKey> const& unlKeys,
+        RCLValidations& validations,
+        std::shared_ptr<SHAMap> const& initialSet);
+
+    /**
+     * Notify NegativeUNLVote that new validators are added.
+     * So that they don't get voted to the NegativeUNL immediately.
+     *
+     * @param seq the current LedgerIndex when adding the new validators
+     * @param nowTrusted the new validators
+     */
+    void
+    newValidators(LedgerIndex seq, hash_set<NodeID> const& nowTrusted);
+
+private:
+    NodeID const myId_;
+    beast::Journal j_;
+    mutable std::mutex mutex_;
+    hash_map<NodeID, LedgerIndex> newValidators_;
+
+    /**
+     * Add a ttUNL_MODIFY Tx to the transaction set.
+     *
+     * @param seq the LedgerIndex when adding the Tx
+     * @param vp the master public key of the validator
+     * @param disabling disabling or re-enabling the validator
+     * @param initialSet the transaction set
+     */
+    void
+    addTx(
+        LedgerIndex seq,
+        PublicKey const& vp,
+        bool disabling,
+        std::shared_ptr<SHAMap> const& initialSet);
+
+    /**
+     * Pick one candidate from a vector of candidates.
+     *
+     * @param randomPadData the data used for picking a candidate.
+     *        @note Nodes must use the same randomPadData for picking the same
+     *        candidate. The hash of the parent ledger is a good choice.
+     * @param candidates the vector of candidates
+     * @return the picked candidate
+     */
+    NodeID
+    pickOneCandidate(
+        uint256 const& randomPadData,
+        std::vector<NodeID> const& candidates);
+
+    /**
+     * Build a reliability measurement score table of validators' validation
+     * messages in the last flag ledger period.
+     *
+     * @param prevLedger the parent ledger
+     * @param unl the trusted master keys
+     * @param validations the validation container
+     * @return the built scoreTable or empty optional if table could not be
+     * built
+     */
+    std::optional<hash_map<NodeID, std::uint32_t>>
+    buildScoreTable(
```

Comment:
This function modified `validations`. I'd rather have an interface that did not modify `validation`, but if that's difficult we at least need to document the face that `validations` is an in/out parameter and how this function changes that parameter.

Reply:
Right, Added documentation.